internal/core/components/watcher.go

```diff
@@ -0,0 +1,182 @@
+package components
+
+import (
+	"context"
+	"runtime"
+	"sync"
+
+	"github.com/api7/apisix-seed/internal/core/comm"
+	"github.com/api7/apisix-seed/internal/core/entity"
+	"github.com/api7/apisix-seed/internal/core/storer"
+	"github.com/api7/apisix-seed/internal/discoverer"
+	"github.com/api7/apisix-seed/internal/utils"
+)
+
+type Watcher struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// Limit the number of simultaneously query
+	sem chan struct{}
+}
+
+func (w *Watcher) Init() {
+	// the number of semaphore is referenced to https://github.com/golang/go/blob/go1.17.1/src/cmd/compile/internal/noder/noder.go#L38
+	w.sem = make(chan struct{}, runtime.GOMAXPROCS(0)+10)
+
+	// List the initial information
+	for _, s := range storer.GetStores() {
+		objPtrs, err := s.List(entity.ServiceFilter)
+		if err != nil {
+			panic("storer list error")
+		}
+
+		wg := sync.WaitGroup{}
+		wg.Add(len(objPtrs))
+		for _, objPtr := range objPtrs {
+			w.sem <- struct{}{}
+			go w.handleQuery(objPtr, s.Typ, &wg)
+		}
+		wg.Wait()
+	}
+}
+
+func (w *Watcher) Watch() {
+	w.ctx, w.cancel = context.WithCancel(context.TODO())
+
+	// Watch for entity updates from Storer
+	for _, s := range storer.GetStores() {
+		go w.handleWatch(s)
+	}
+}
+
+func (w *Watcher) Close() {
+	w.cancel()
+
+	for _, s := range storer.GetStores() {
+		s.Unwatch()
+	}
+}
+
+func (w *Watcher) handleQuery(objPtr interface{}, typ string, wg *sync.WaitGroup) {
+	defer func() {
+		<-w.sem
+		wg.Done()
+	}()
+
+	queer := objPtr.(entity.Queer)
+	query, err := encodeQuery(utils.EventAdd, typ, queer)
+	if err != nil {
+		return
```

Comment:
When a function consumes the err, it should log the err down. Please fix other similar places.